### PR TITLE
Support for multiple batches per optimizer step

### PR DESCRIPTION
**Pull request by TortillasAlfred**
With this PR, one can now compute a running loss on multiple batches before proceeding with a weight upgrade, essentially allowing to train a NN on a batch size too big to fit within a GPU's memory by simply fitting multiple smaller batches before upgrading the weights.

Please note that, according to a given parameter `batches_per_step`, the current implementation will divide the running loss of all `batches_per_step` by `batches_per_step`, making the total loss used for the backpropagation an average of the losses on the desired previous batches. This only makes sense in case of losses which are already averaged along the batch size, instead of summed per example. However, since most losses are averaged by default over a given batch, I believe it makes sense to use this out of the box.

### CODE REVIEW

File: poutyne/framework/experiment.py [hunk 1]
```diff
@@ -239,7 +239,9 @@ def train(self,
               epochs=1000,
               steps_per_epoch=None,
               validation_steps=None,
-              seed=42):
+              seed=42,
+              batches_per_step=1):
```
Comment (davebulaval) [inline poutyne/framework/experiment.py]: Whe have a default value too for the constructor. We should only define one default.

File: poutyne/framework/iterators.py [hunk 2]
```diff
@@ -27,11 +27,13 @@ def _get_step_iterator(steps, generator):
 
 
 class StepIterator:
-    def __init__(self, generator, steps_per_epoch, callback, metrics_names):
+    def __init__(self, generator, steps_per_epoch, callback, metrics_names, batches_per_step=1):
```
Comment (davebulaval) [inline poutyne/framework/iterators.py]: see comment before.

File: poutyne/framework/iterators.py [hunk 3]
```diff
@@ -81,7 +83,8 @@ def __init__(self,
                  validation_steps,
                  initial_epoch=1,
                  callback,
-                 metrics_names):
+                 metrics_names,
+                 batches_per_step=1):
```
Comment (davebulaval) [inline poutyne/framework/iterators.py]: See comment before.

File: poutyne/framework/iterators.py [hunk 4]
```diff
@@ -90,6 +93,7 @@ def __init__(self,
         self.initial_epoch = initial_epoch
         self.callback = callback
         self.metrics_names = metrics_names
+        self.batches_per_step = batches_per_step
```
Comment (davebulaval) [inline poutyne/framework/iterators.py]: Do we really need to keep this value ?

File: poutyne/framework/experiment.py (hunk 1 above)
Comment (davebulaval) [reply poutyne/framework/experiment.py]: I think, the model should have the default value not the iterator.

File: poutyne/framework/model.py [hunk 5]
```diff
@@ -334,17 +355,30 @@ def fit_generator(self,
...
+                   batches_per_step=1):
+        if zero_all_gradients:
+            self.optimizer.zero_grad()
 
         loss_tensor, metrics, pred_y = self._compute_loss_and_metrics(x,
                                                                       y,
                                                                       return_loss_tensor=True,
                                                                       return_pred=return_pred)
 
+        loss_tensor = loss_tensor / batches_per_step
```
Comment (davebulaval) [inline poutyne/framework/model.py]: do we need to validate a negative number ? or a too large batches per step value ?

File: poutyne/framework/experiment.py (hunk 1 above)
Comment (TortillasAlfred) [reply poutyne/framework/experiment.py]: Not sure what you mean here: default value has been set to 1 for `batches_per_step` because the default use-case of the whole library is not what this particular PR aims to solve.

File: poutyne/framework/iterators.py (hunk 2 above)
Comment (TortillasAlfred) [reply poutyne/framework/iterators.py]: Was a mistake. Fixed by 874c3f709364a90f3c5b28faf6118b27fd3185f2

File: poutyne/framework/iterators.py (hunk 3 above)
Comment (TortillasAlfred) [reply poutyne/framework/iterators.py]: Also a mistake. Fixed by 874c3f709364a90f3c5b28faf6118b27fd3185f2

File: poutyne/framework/iterators.py (hunk 4 above)
Comment (TortillasAlfred) [reply poutyne/framework/iterators.py]: Mistake. Fixed by 874c3f709364a90f3c5b28faf6118b27fd3185f2

File: poutyne/framework/model.py (hunk 5 above)
Comment (TortillasAlfred) [reply poutyne/framework/model.py]: I know many libraries like Numpy and Scikit-learn do some kind of sanity check on all input arguments to provide their users with potentially more meaningful error messages. 

However, since this is not seen anywhere in Poutyne, I don't feel like it's needed to go beyond a clear description of the parameter's role, hence its possible values.

File: poutyne/framework/experiment.py (hunk 1 above)
Comment (freud14) [reply poutyne/framework/experiment.py]: I think this is fine. However, maybe put the `batches_per_step` argument before the `seed` argument since the `seed` argument is less important and fits more with the previous arguments.

Comment (freud14) [reply poutyne/framework/experiment.py]: And do the same for every other places. I think we should put it after the `validation_steps` argument at every places.

File: poutyne/framework/model.py [hunk 6]
```diff
@@ -334,17 +354,30 @@ def fit_generator(self,
...
```
Comment (freud14) [inline poutyne/framework/model.py]: I don't think that the loss that is sent in the logs should be divided. It should only be divided for the backward pass.

Comment (freud14) [reply poutyne/framework/model.py]: Also, if we want to support uneven batches, the loss should be multiplied by the batch size. Then, the gradient should be divided by the total number of examples for the optimizer step.

File: poutyne/framework/model.py (hunk 5 above)
Comment (davebulaval) [reply poutyne/framework/model.py]: Good with that.

File: poutyne/framework/model.py (hunk 6 above)
Comment (davebulaval) [reply poutyne/framework/model.py]: I think the same.

File: poutyne/framework/experiment.py (hunk 1 above)
Comment (TortillasAlfred) [reply poutyne/framework/experiment.py]: Fixed by fecbe56534bf75d572f69f4c09a537505c4e8041

File: poutyne/framework/model.py (hunk 6 above)
Comment (TortillasAlfred) [reply poutyne/framework/model.py]: Fixed by ec1d849831f6d02afc8c658098c0f5db20f4a65e.

File: poutyne/framework/experiment.py [hunk 7]
```diff
@@ -298,7 +300,8 @@ def train(self,
                                             steps_per_epoch=steps_per_epoch,
                                             validation_steps=validation_steps,
                                             initial_epoch=initial_epoch,
-                                            callbacks=callbacks)
+                                            callbacks=callbacks,
+                                            batches_per_step=batches_per_step)
```
Comment (freud14) [inline poutyne/framework/experiment.py]: Put this after validation_steps.

File: poutyne/framework/iterators.py [hunk 8]
```diff
@@ -28,6 +28,7 @@ def _get_step_iterator(steps, generator):
 
 class StepIterator:
     def __init__(self, generator, steps_per_epoch, callback, metrics_names):
+        # pylint: disable=too-many-arguments
```
Comment (freud14) [inline poutyne/framework/iterators.py]: Don't think it's necessary since pylint was passing before.

File: poutyne/framework/model.py [hunk 9]
```diff
@@ -220,7 +225,8 @@ def fit(self,
                                   validation_steps=validation_steps,
                                   initial_epoch=initial_epoch,
                                   verbose=verbose,
-                                  callbacks=callbacks)
+                                  callbacks=callbacks,
+                                  batches_per_step=batches_per_step)
```
Comment (freud14) [inline poutyne/framework/model.py]: Put this after validation_steps.

File: poutyne/framework/model.py [hunk 10]
```diff
@@ -323,32 +334,77 @@ def fit_generator(self,
...
+                    examples_in_step += step.size
+
+                    step.loss, step.metrics, _ = self._fit_batch(x,
+                                                                 y,
+                                                                 callback=callback_list,
+                                                                 step=step.number,
+                                                                 zero_all_gradients=zero_all_gradients,
+                                                                 do_backprop=do_backprop,
+                                                                 batch_size=step.size,
+                                                                 examples_in_step=examples_in_step)
```
Comment (freud14) [inline poutyne/framework/model.py]: I think that batch_size and examples_in_step should be None when batches_per_step is 1 so that the gradient and loss are not modified for nothing in that case.
